docs(bodiless-documentation): Design Elements Basics tutorial (#147)
johnsonandjohnson/Bodiless-JS: packages/bodiless-documentation/doc/About/DesignElementBasics.md

# Design Element Basics

> In this guide, we will continue to the tutorial of the gallery and apply the some designs & best practices of site building.

Author: Addressed

## 1. Create a Site

> [Create a new site](https://johnsonandjohnson.github.io/Bodiless-JS/#/About/GettingStarted?id=creating-a-new-site) and launch the editor interface.

Author: Resolved

> ## 2. Copy Gallery example to Site & Use Site's Simple Editor

Author: Resolved

> This step we are going to copy over the tutorial that was created in the [Site Build Basics guide](https://johnsonandjohnson.github.io/Bodiless-JS/#/About/SiteBuildBasics) and having it use the site's predefined rich text editors that are provided by starter kit. There is no need to use simple editor defined at page level in the initial tutorial and its best practice to define rich text editors a site uses at component level.

Author: Resolved

For the Best Practices -- I think should be part of the Site build tutorial and not here.

1. Copy over the [gallery-final folder & contents](https://github.com/johnsonandjohnson/Bodiless-JS/tree/master/examples/test-site/src/data/pages/gallery-final) and place in site at `src/data/pages/gallery`  Note: remember to rename the folder from gallery-final to gallery.

> 1. Lets replace the SimpleEditor with Site's Basic editor as there is no need to use specific editor at page level. 

Author: Resolved

   * Replace `import withSimpleEditor from './withSimpleEditor';`  
     with `import { asEditorBasic } from '../../../components/Editors';`
   * Replace `const Body = withSimpleEditor('body', 'Body')(Fragment);`
     with `const Body = asEditorBasic('body', 'Body')(Fragment);`
   * Repeat above steps in `CaptionedImage.tsx`
   * Delete the file `withSimpleEditor.tsx`
1. Run your site and visit the gallery page (http://localhost:8000/gallery) and it should run exactly like it did in examples/test site with slightly different choices for the Rich Text Editor in the body of the gallery page. 

## 3. Create the Gallery as reusable component to be used within site.

> If you are going to have component reusable for a page or template, its best to have them placed in `src/components` so they can be reused by any page/template.  While you could import them from another page, it's not best practice and makes future finding and maintaining hard finding components within a page.

Author: Resolved

1. Create a folder call Gallery in `src/components`
1. Move the Gallery.tsx & CaptionedImage.tsx to the `/src/components/Gallery` folder
1. Rename Gallery.tsx -> index.tsx 
1. Change the import on the page to import Gallery & GalleryTile from `../../../components/Gallery`
1. Run your site and visit the gallery page (http://localhost:8000/gallery) and it should run exactly like it did in examples/test site. 

## 4. Create a reuseable Primary Header for the site 

> Within `data/pages/gallery/index.jsx` (gallery page) & `data/pages/index.jsx` (homepage) you can see we use the similar primary L1 header.  Let's move the primary header h1 to be shared. This way if we change the style of h1 primary header it will apply throughout the site instead of having to fix each page. 

Author: Resolved

1. In `src/componets/Elements.token.ts` lets define a new H1 Primary Header
    ```
    const PrimaryHeader = flow(
      asHeader1,
      asBold,
      asEditable('title', 'Page Title'),
    )(H1);
    ````
1. Remember to add imports needed & export this new component.
1. Change the imports on both gallery index.jsx & homepage page index.jsx and remove the current PrimaryHeader definitions.
1. Run your site and visit the homepage & gallery page (http://localhost:8000/gallery) and it should run exactly like it did in example site except the gallery title is not bold. 
1. In `src/componets/Elements.token.ts` add a tailwind to class to asBold. 
    ```
    const asBold = addClasses('font-bold');
    ```

> 1. Visit the homepage & gallery page (http://localhost:8000/gallery) and both H1 titles should be bold.

Author: Resolved

## 5.  Create a reuseable Primary Header Image for the site 

> The header image is also a case where we could create a more flexible linkable image header.  Let's go ahead and create a component that we can reuse and apply the design api to make some variations.

Author: Resolved

* The header on the homepage is a simple full width image.
* The header on the gallery page is linkable image. 

> Lets create a HeaderImage that is fill width linkable image with a stylized token that gives option to remove the link.

Author: Resolved

1. Create HeaderImage folder

> Let's create a HeaderImage folder in `/src/components` and create two files within this folder `index.tsx` & `token.tsx`.  This is a best practice to start creating components in this manner.  The index.tsx contains your definition of what your component does and token.tsx contains definitions of how your component looks.

Author: @e2thex  @wodenx 
Changed to "The index.tsx contains the template defining how your component functions and token.tsx contains the design elements control how your component looks."

1. Lets start work on the `index.tsx` and get the imports setup.  This is pretty self explanatory. If you forget one you will be warned and it won't work!
    ```
    import React from 'react';
    import { flow } from 'lodash';
    import {
      withDesign,
      designable,
      DesignableComponentsProps,
      Div,
      A,
      Img,
      StylableProps,
    } from '@bodiless/fclasses';
    import {
      asBodilessImage,
      asBodilessLink,
    } from '@bodiless/components';
    import { withNode } from '@bodiless/core';
    ```

> 1. The first step is to setup that every individual component is stylable and then define what that component is.  This will allow us to replace or modify the individual components within this component. 

Author: Resolve

   * The HeaderImageComponents define each individual component as Styleable Props. 
   * The headerImageStart:HeaderImageComponents defines what each component is. 
    ```
    export type HeaderImageComponents = {
      ImageWrapper: ComponentType<StylableProps>,
      ImageLink: ComponentType<StylableProps>,
      Image: ComponentType<StylableProps>,
    };

    const headerImageStart:HeaderImageComponents = {
      ImageWrapper: Div,
      ImageLink: A,
      Image: Img,
    };

>     type Props = DesignableComponentsProps<ToutComponents> & { };

Author: Resolved

    ```

> 1. Next, lets generate how the component will render. It will generate an linked image in a div wrapper.

Author: Resolved

    ```
    const HeaderImageBase: FC<Props> = ({ components }) => {
      const {
        ImageWrapper,
        Image,
        ImageLink,
      } = components;

      return (
        <ImageWrapper>
          <ImageLink>
            <Image />
          </ImageLink>
        </ImageWrapper>
      );
    };

>     ```

Author: Resolved

> 1. Now, lets combine the individual elements we have with the designable interface so they can take the Design prop. We are also wrapping it with withNode() to ensure that all its children have a unique location to store their data: 

Author: Resolved

    ```
    const HeaderImageClean = flow(
      designable(headerImageStart),
      withNode,
    )(HeaderImageBase);

> 

Reviewer: 1. Next, lets pass in some HOC's via `withDesign` to make our component editable.

Author: Resolved

    const asEditableHeaderImage = flow(
      withDesign({
        Image: asBodilessImage('headerimage'),
        ImageLink: asBodilessLink('imagelink'),
      }),
    );

>     ```

Reviewer: `withDesign` takes an object whose keys are the names of the subcomponents which belong to our `HeaderImage`, and whose values are higher-order components which should be applied to each.  As you'll see below, these HOC's often carry elements of styling, and correspond to the "tokens" of a design system.  Here, they apply functionality, but the pattern is the same.

Note that these HOC's (or tokens) are defined at the site level.  For now, they are just pass-through's to the core *Bodiless* utilities - but in manuy cases you will want to customize these at the site level (for example, to provide a different image selector, or a link with an editable `target` attribute).

```

Author: Resolved

1.  Lastly, lets combine these together and export. 
    ```
    const HeaderImage = asEditableHeaderImage(HeaderImageClean);

>     export default HeaderImage;

Reviewer: Should we be exporting the component here or the asEditableHeaderImage() HOC?  cc @e2thex

    ```
    

> 1. Turning our focus to the `token.tsx`, this is where we will define what the component will look like and abstract the design here.  We start with the imports and then export to 2 different designs styles with:

Reviewer: ```suggestion
1. Turning our focus to the `token.tsx`, this is where we will define some variations on what the component will look like .  We start with the imports and then export two different stylings, with
```

Author: resolved

> * asHeaderImageDefaultStyle being a full width image, with rounded corners

Reviewer: ```suggestion
* `asHeaderImageDefaultStyle` being a full width image, with rounded corners
```

Author: Resolved

> * AsHeaderImageNotLinkable simply removes the Linkable image.

Reviewer: ```suggestion
* `asHeaderImageNotLinkable` simply removing the link from the image.
```

Reviewer: Note that these HOC's themselves can be considered "tokens" which describe design elements which can be applied to our `HeaderImage` as a whole.

Author: Resolved

    ```
    import React from 'react';
    import { flow } from 'lodash';
    import {
      withDesign,
      addClasses,
      remove,
    } from '@bodiless/fclasses';
    import { asImageRounded } from '../Elements.token';

    export const asHeaderImageDefaultStyle = withDesign({
      ImageWrapper: addClasses('relative'),
      ImageLink: addClasses(''),
      Image: flow(
        addClasses('m-0 w-full h-auto'),
        asImageRounded,
      ),
    });

>     export const asHeaderImageNotLinkable = withDesign({

Reviewer: I'm not sure about this.  Would the correct pattern be instead to create a HeaderImage component without the link, and then an `asLinkableHeaderImage` HOC which adds the link (by wrapping the Image with a Link)?  OTOH, this displays how you can take an existing component and customize it (by removing something).  And we use this pattern in Touts, right?  Rather than making the bare bones tout and building up the ones with more stuff?  Why so?  cc @e2thex

      ImageLink: remove,
    });

    ```

> 1. We also need a new token of rounded corners that would be useful to a token, so place this `Elements.token.ts` file:

Reviewer: ```suggestion
1. Note above we used  a new token describing rounded corners (`asImageRounded`).  This would belong to the design system of the site, and should be placed in an `Elements.token.ts` file:
```

Author: Resolved

`const asImageRounded = addClasses('rounded-lg');`

> and remember to export it.

Reviewer: Do we want something here describing how we're using fclasses to compose the tokens?

Author: Resolved

## 6. Let's use that Header Image Component both on the front page and gallery page.

> 1. On the homepage, import in the HeaderImage and styled Headers and remove the previous local HeaderImage definition:

Reviewer: ```suggestion
1. On the homepage, import in the `HeaderImage` and tokens, and remove the previous local HeaderImage definition:
```

Author: Resolved

    ```
    import HeaderImage from '../../components/HeaderImage';
    import { asHeaderImageDefaultStyle, asHeaderImageNoLink } from '../../components/HeaderImage/token';

>     const HomePageHeaderImage = flow(asHeaderImageDefaultStyle, asHeaderImageNoLink)(HeaderImage);

Reviewer: See - I think we should be applying `asEditableHeaderImage`here, rather than exporting the editable version.  This makes it easier to place a non-editable header image using all he other tokens.

    ```

    and replace

    ```
      <div className="flex my-3 w-full">
        <HeaderImage />
      </div>
    ```
    
    with ```<HomePageHeaderImage />```

> 2. On the gallery, import in the HeaderImage and styled Headers and remove the previous local HeaderImage definition:

Reviewer: ```suggestion
2. On the gallery, import in the `HeaderImage` and tokens, and remove the previous local `HeaderImage` definition:
```

Author: Resolved

    ```
    import HeaderImage from '../../../components/HeaderImage';
    import { asHeaderImageDefaultStyle } from '../../../components/HeaderImage/token';

    const GalleryHeaderImage = asHeaderImageDefaultStyle(HeaderImage);
    ```

    and replace 

    ```
      <div className="flex my-3 w-full">
        <HeaderImage />
      </div>
    ```
    
    with ```<GalleryHeaderImage />```

1. Visit the homepage & gallery page (http://localhost:8000 & http://localhost:8000/gallery) and homepage should be full width image and gallery should be linkable image.

> While this is simple component, imagine that a request came in that all images headers on all pages, should not have rounded corners.  You delete one element token out of asHeaderImageDefaultStyle and it takes effect on any Header Image in use on any page using this style and doesn't require going through each page.

Reviewer: Is it worth a paragraph here describing how this is better than (and doesn't just replicate) the normal CSS cascade?

Author: Resolved

          Image: ComponentType<StylableProps>,
          Body: ComponentType<StylableProps>,
        };
        ```
    
   * Then we define what to render by default for each component; we use the
     stylable versions of basic HTML elements exported by
     [`@bodiless/fclasses`](https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/fclasses/src/StyledHTML.tsx).
        ```
        const captionedImageStart:CaptionedImageComponents = {
          Wrapper: Section,
          Image: Img,
          Body: Div,
        };

        type Props = DesignableComponentsProps<CaptionedImageComponents> & { };
        ```

1. Now that we have our sub-components, let's define how they go together to
   make a captioned image. This is really the base template for our component.
   Combining the defaults defined above with this layout, we will render an
   `img` & `div` wrapped in a `section`.
    ```
    const CaptionedImageBase: FC<Props> = ({ components }) => {
      const {
        Wrapper,
        Image,
        Body,
      } = components;

      return (
        <Wrapper>
          <Image />
          <Body />
        </Wrapper>
      );
    };
    ```
    

>     Note: that the actual sub-components here are *injected*; that is, they are passed into the component via a `components` prop. We defined the defaults for these components above (`captionedImageStart`), but we will actually render whatever we are passed.

Reviewer: ```suggestion
    Note: that the actual sub-components here are *injected*; that is, they
    are passed into the component via a `components` prop. We defined
    the defaults for these components above (`captionedImageStart`), but
    we will actually render whatever we are passed.
```

    contexts where we don't want to allow the content to be edited.
  * `asEditableCaptionedImage` is a *component level* token which makes our
    clean component editable, and can be applied wherever we do want to allow
    the content to be edited.
  * `CaptionedImage` combines the above and is exported for convenience, since
    we expect that in most cases we'll want the component to be editable.

  Note, that we have not, as yet, applied any styling or design tokens.

1. As a final step in this file, we make sure the imports are correct. This is
   pretty self explanatory. If you forget one you will be warned and it won't
   work!
    ```
    import React, { FC } from 'react';
    import { asBodilessImage } from '@bodiless/components';
    import { withNode } from '@bodiless/core';
    import {
      withDesign,
      designable,
      DesignableComponentsProps,
      Div,
      Section,
      Img,
      StylableProps,
    } from '@bodiless/fclasses';
    import { flow } from 'lodash';
    import { asEditorBasic } from '../Editors';
    ```

## 5. Combine `CaptionedImage` & `PrimaryHeader` on the homepage.

> Let's take the components we have just created and combine them. Imagine that our design calls for a page header block on the homepage with image and a header text.

Reviewer: ```suggestion
Let's take the components we have just created and combine them. Imagine
that our design calls for a page header block on the homepage with image and
a header text.
```

1. On the homepage, import the `CaptionedImage` component and define this new variation:

    ```
    import { CaptionedImage } from '../../components/Gallery-Design/CaptionedImage';
    
    const PageHeader = flow(
      withDesign({
        Body: replaceWith(PrimaryHeader),
      }),
    )(CaptionedImage);
    ```

>     You can see we "modified the design" of the original captioned image, replacing the body (previously defined as a `div` wrapped in `asBasicEditor`) with our `PrimaryHeader.

Reviewer: ```suggestion
    You can see we "modified the design" of the original captioned image, replacing
    the body (previously defined as a `div` wrapped in `asBasicEditor`) with our
    `PrimaryHeader.
```

>     Note the use of the `replaceWith` HOC here.  Remember that the object passed to `withDesign` is always a set of higher-order components.  In most cases you will just want to use these to apply styling to the default elements defined in the component's design.  For example, we might have written:

Reviewer: ```suggestion
    Note the use of the `replaceWith` HOC here.  Remember that the object
    passed to `withDesign` is always a set of higher-order components.  In most
    cases you will just want to use these to apply styling to the default elements
    defined in the component's design.  For example, we might have written:
```

    ```
    const PageHeader = flow(
       withDesign({
        Body: asHeaderPrimary,
      }),
    )(CaptionedImage);
    ```
    Here, however, the default element is a `div`, while we want a `h1`, so we have to replace it.
    
    On the homepage remove the combination of HeaderImage & PrimaryHeader

    ```
      <div className="flex my-3 w-full">
        <HeaderImage />
      </div>
      <PrimaryHeader />
    ```
    
    and replace with 
    
    ```
      <PageHeader />
    ```
    

>     Reload the homepage and make sure it renders as expected. We could take this step farther and move our `HeaderImage` to `src/components` and reuse it across all the pages. It could also have linkable header image, apply the title over the image, etc. We will leave this as exercise for you to do on your own.

Reviewer: ```suggestion
    Reload the homepage and make sure it renders as expected. We could take
    this a step farther and move our `HeaderImage` to `src/components` and reuse
    it across all the pages. It could also have linkable header image, apply the title
    over the image, etc. We will leave this as exercise for you to do on your own.
```

> While this a simple component we are wrapping in the design, proceeding in this manner as the components grow in either functionality or complexity gives us a few benefits:

Reviewer: ```suggestion
While this a simple component we are wrapping in the design, proceeding in this
manner as the components grow in either functionality or complexity gives us a
few benefits:
```

* Design is separated from the internal markup of the component.

> * Simplified Styling: this simplifies styling of components and eliminates the normal css cascade that builds and grows over time.

Reviewer: ```suggestion
* Simplified Styling: this simplifies styling of components and eliminates the normal
css cascade that builds and grows over time.
```

> * Isolation: it keeps the styling isolated to the specific item minimizing the risk of affecting other non-related items.

Reviewer: ```suggestion
* Isolation: it keeps the styling isolated to the specific item minimizing the risk
of affecting other non-related items.
```

* Reuseability as is or with extending.

> These benefits apply during the initial build and future changes benefit as well. For example, if there is a request to change a rendered H1 to H2 for SEO purposes, this can easily be achieved.

Reviewer: ```suggestion
These benefits apply during the initial build and future changes benefit as well.
For example, if there is a request to change a rendered H1 to H2 for SEO purposes,
this can easily be achieved.
```

## 6. Continue with wrapping the `Gallery` with Design API

> 1. The current files in Gallery folder contain the templates defining how your component functions. Let's create a `token.tsx` and move the design styles to a separate file.

Reviewer: ```suggestion
1. The current files in Gallery folder contain the templates defining how
your component functions. Let's create a `token.tsx` and move the
design styles to a separate file.
```

    ```
    import {
      withDesign,
      addClasses,
    } from '@bodiless/fclasses';

    const asGalleryDefaultStyle = withDesign({
      Wrapper: addClasses('my-2'),
      Header: addClasses('text-2xl'),
    });

    const asImageTile = withDesign({
      Wrapper: addClasses('mx-2 border-8'),
      Image: addClasses('w-full'),
    });

    const asBlueBorder = withDesign({
        Wrapper: addClasses('border-blue-400'),
    });

    const asRedBorder = withDesign({
      Wrapper: addClasses('border-red-400'),
    });
    const asGreenBorder = withDesign({
        Wrapper: addClasses('border-green-400'),
    });

    export {
      asImageTile,
      asBlueBorder,
      asRedBorder,
      asGreenBorder,
      asGalleryDefaultStyle,
    }
    ```
    

>     These HOC's themselves can be considered "Component Tokens" which describe design elements which can be applied to the components as a whole. In other words, "Component Tokens" are no different than normal Element tokens except they apply to multiple sub-components. In `asImageTile`, you can see we added margin, border to the wrapper and made sure the image shows full-width. All tokens here were taken from the existing `Gallery/index.tsx` file.

Reviewer: ```suggestion
    These HOC's themselves can be considered "Component Tokens" which describe
    design elements which can be applied to the components as a whole. In other
    words, "Component Tokens" are no different than normal Element tokens except
    they apply to multiple sub-components. In `asImageTile`, you can see we added
    margin, border to the wrapper and made sure the image shows full-width. All tokens
    here were taken from the existing `Gallery/index.tsx` file.
```

    
    **TIP**: By convention all Component Tokens start with `as`.

> 1. Let's update the `Gallery/index.tsx` and use the component tokens in place of the current styling. In addition, let's wrap the Gallery Component in the Design API as well, using the same method we just did.

Reviewer: ```suggestion
1. Let's update the `Gallery/index.tsx` and use the component tokens in place
of the current styling. In addition, let's wrap the Gallery Component in the
Design API as well, using the same method we just did.
```
